### PR TITLE
feat: add method to add metadata section values as a dictionary (restarted)

### PR DESCRIPTION
**Pull request by robinmacharg**
[Note: this PR replaces #456 ([Branch here](1c0b988f53d907085e9fb6d383c426ec3c13624b)) which was proving difficult to merge.  The PR text is largely unchanged.]

## Goal

Adds a `BugsnagMetadata.addMetadataToSection:values:` method to allow merging arbitrary dictionaries with preexisting Metadata.

## Design

Simple fast enumeration loop over provided values with appropriate checks for:

- Validity of Key (String subclass)
- Validity of Value (can be sanitised: a recursive check that all (nested) values are of type Number, String, Array, Dictionary or Null

Conditional addition/deletion depending on whether value is null or not.  There is a check to ensure that empty sections are not created for (failing) invalid values.  Invalid values are logged and ignored.  Delegate callback is called only where there is a material - CRUD - change to the metadata.

## Changeset

- Method added to `BugsnagMetadata`.
- Unit Tests added.

## Tests

- Added Unit tests only.  This change affects the framework API and doesn't change any external aspect.
- CI passes.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [X] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [X] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [X] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,6 @@
 Changelog
 =========
 
-## TBD
-
```
Comment (kattrali) [inline CHANGELOG.md]: No need to remove this. In fact, its [automatically replaced during the release process with the new version number and date](https://github.com/bugsnag/bugsnag-cocoa/blob/e95c825c64cb359ba87dfec32dfa66a11ce47dc9/Makefile#L67).

File: CHANGELOG.md [hunk 2]
```diff
@@ -78,6 +76,10 @@ Bugsnag Notifiers on other platforms.
 * Add metadata accessor methods to `BugsnagEvent`
   [#465](https://github.com/bugsnag/bugsnag-cocoa/pull/465)
 
+* Added a method allowing merging arbitrary (JSON-encodable) dictionaries with 
+  an existing Event Metadata instance (`BugsnagMetadata.addMetadataToSection:values:`)
```
Comment (kattrali) [inline CHANGELOG.md]: What does this mean for users exactly? Saying that it "merges" dicts with Event Metadata could be unclear about what will happen when this method is called. Can you provide a code example here? The headerdoc for the function is fairly clear.

File: Source/BugsnagMetadata.m [hunk 3]
```diff
@@ -94,33 +94,107 @@ - (NSDictionary *)toDictionary {
...
             id cleanedValue = BSGSanitizeObject(value);
             if (cleanedValue) {
-                NSDictionary *section = [self getMetadata:sectionName];
+                // Value is OK, try and set it
+                NSMutableDictionary *section = [self getMetadata:tabName];
                 if (!section) {
-                    section = [NSMutableDictionary new];
-                    [[self dictionary] setObject:section forKey:sectionName];
+                    section = [@{} mutableCopy];
+                    [[self dictionary] setObject:section forKey:tabName];
```
Comment (kattrali) [inline Source/BugsnagMetadata.m]: Not necessary bad, but why the change from making a dictionary using the constructor to using a copy?

File: CHANGELOG.md (hunk 1 above)
Comment (robinmacharg) [reply CHANGELOG.md]: TIL, thanks.  Added back in.

File: CHANGELOG.md (hunk 2 above)
Comment (robinmacharg) [reply CHANGELOG.md]: I've expanded the message to have what I hope is enough clarity.  If you still think it would benefit from a code sample let me know.

File: Source/BugsnagMetadata.m (hunk 3 above)
Comment (robinmacharg) [reply Source/BugsnagMetadata.m]: I'm not actually sure why I made the change.  I agree, though, why jump through an unnecessary hoop?  Changed back.
